include/seqan3/index/fm_index.hpp

```diff
@@ -0,0 +1,203 @@
+ *     std::cout << "Positions in the genome: " << it.locate() << '\n'; // outputs: TODO
+ *
+ *     return 0;
+ * }
+ * ```
+ *
+ * Even though the FM index is originally a prefix tree and one performs backward searches, for convenience
+ * of the user it is implemented as a suffix tree. There is no need to reverse the text to be indexed, the patterns
+ * to be searched or recompute positions.
+ *
+ * Here is an example using a collection of strings (e.g. a genome with multiple chromosomes or a protein database):
+ *
+ * TODO
+ *
+ * There is also a history iterator, i.e. an iterator that stores its previous states on a stack such that going down
+ * an edge can be undone. Please take a look at the documentation of seqan3::fm_index_history_iterator::up() since it
+ * does not undo all operations.
+ *
+ * ### Choosing an index implementation
+ *
+ * The underlying implementation of the FM Index (Rank data structure, sampling rates, etc.) can be specified ...
+ */
+template <input_range_concept text_t, fm_index_traits_concept fm_index_traits = fm_index_default_traits>
+    requires alphabet_concept<innermost_value_type_t<text_t>>
+class fm_index
+{
+protected:
+    using sdsl_index_type = typename fm_index_traits::sdsl_index_type;
+    using comp_char_type = typename sdsl_index_type::alphabet_type::comp_char_type;
+
+    sdsl_index_type m_index;
+    text_t const * text = nullptr;
+
+public:
+    using text_type = text_t;
+    using char_type = innermost_value_type_t<text_t>;
+    using size_type = typename sdsl_index_type::size_type;
+
+    using iterator_type = fm_index_iterator<fm_index<text_t, fm_index_traits>>;
+    friend class fm_index_iterator<fm_index<text_t, fm_index_traits>>;
+    friend class detail::fm_index_iterator_node<fm_index<text_t, fm_index_traits>>;
+
+    // static const bool is_bidirectional = false;
+    // static constexpr uint8_t dimensions = DIMENSIONS;
+
+    fm_index() = default;
+    fm_index(fm_index const &) = default;
+    fm_index & operator=(fm_index const &) = default;
+    fm_index(fm_index &&) = default;
+    fm_index & operator=(fm_index &&) = default;
+
+    // TODO: requires common_reference_concept<detail::innermost_value_type_t<container_t>, char_t> && detail::dimension_v<container_T> == dimensions
+    fm_index(text_t const & text)
+    {
+        construct(text);
+    }
+
+    void construct(text_t const & text)
+    {
+        this->text = &text;
+        assert(!text.empty());
+        // TODO: check what happens in sdsl when constructed twice!
+        // TODO: choose between in-memory/external and construction algorithms
+        // TODO: sdsl construction currently only works for int_vector, std::string and char *, not ranges in general
+        sdsl::int_vector<8> tmp_text(text.size());
+        for (auto it = text.cbegin(); it != text.cend(); it++)
+            tmp_text[text.cend() - it - 1] = to_rank(*it) + 1; // reverse and increase rank by one
```

Comment:
dna4 is only included in the concepts.
true ... the user has to include an alphabet type anyway to load/create a container of that alphabet anyway. but the user might not include the alphabet module before the index module :bulb:

Reply:
Yeah, when i thought about it, it also occurred to me, that it may be hard to use an alphabet and pass it to the fm_index without including the alphabet - so it's enough to include 1 alphabet :)